src/alignment.js

```diff
@@ -90,15 +90,22 @@ var alignWithDOM = function(nodeName, key, statics) {
       if (process.env.NODE_ENV !== 'production') {
         assertKeyedTagMatches(existingNode, nodeName, key);
       }
+
       matchingNode = existingNode;
     } else {
       matchingNode = createNode(walker.doc, nodeName, key, statics);
+
       if (key) {
         registerChild(parent, key, matchingNode);
       }
     }
 
-    parent.insertBefore(matchingNode, currentNode);
+    if (currentNode && getData(currentNode).key) {
+      parent.replaceChild(matchingNode, currentNode);
```

Comment:
Hmm, we could add a flag to the keyMap to mark it as invalid and recreate it on the next pass. Or perhaps instead of completely recreating it, go through the keyMap and see what is still attached to the DOM and remove those that aren't.